src/vcRender.cpp

```diff
@@ -665,6 +711,69 @@ void vcRender_VisualizationPass(vcState *pProgramState, vcRenderContext *pRender
   vcMesh_Render(gInternalMeshes[vcInternalMeshType_ScreenQuad]);
 }
 
+void vcRender_ApplyViewShed(vcRenderContext *pRenderContext)
+{
+  vcShader_Bind(pRenderContext->shadowShader.pProgram);
+  vcShader_BindTexture(pRenderContext->shadowShader.pProgram, pRenderContext->pDepthTexture[0], 0, pRenderContext->shadowShader.uniform_depth);
+  vcShader_BindTexture(pRenderContext->shadowShader.pProgram, pRenderContext->viewShedRenderingContext.pDepthTex, 1, pRenderContext->shadowShader.uniform_shadowMapAtlas);
+
+  vcShader_BindConstantBuffer(pRenderContext->udRenderContext.presentShader.pProgram, pRenderContext->shadowShader.uniform_params, &pRenderContext->shadowShader.params, sizeof(pRenderContext->shadowShader.params));
+
+  vcMesh_Render(gInternalMeshes[vcInternalMeshType_ScreenQuad]);
+}
+
+void vcRender_RenderAndApplyViewSheds(vcState *pProgramState, vcRenderContext *pRenderContext, vcRenderData &renderData)
+{
+  if (renderData.viewSheds.length == 0)
+    return;
+
+  vcGLState_SetDepthStencilMode(vcGLSDM_Always, false);
+  vcGLState_SetBlendMode(vcGLSBM_Additive);
+
+  pRenderContext->shadowShader.params.inverseProjection = udFloat4x4::create(udInverse(pProgramState->pCamera->matrices.projection));
+
+  if (pRenderContext->viewShedRenderingContext.pRenderView == nullptr)
+    vdkRenderView_Create(pProgramState->pVDKContext, &pRenderContext->viewShedRenderingContext.pRenderView, pRenderContext->udRenderContext.pRenderer, ViewShedMapRes.x / ViewShedMapCount, ViewShedMapRes.y);
+
+  for (size_t v = 0; v < renderData.viewSheds.length; ++v)
```

Comment:
Any plans to cache these to avoid the UD rerender/uploads if not dirty?

Reply:
Task added.
AB#437